x/archivedb/key.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package archivedb
+
+import (
+	"encoding/binary"
+	"errors"
+	"math"
+)
+
+var (
+	ErrInsufficientLength = errors.New("packer has insufficient length for input")
+	longLen               = 8
+	boolLen               = 1
+	// This default prefix is used as the internal default prefix for all keys.
+	// By doing so the database can safely accept keys from untrusted sources,
+	// by using a common prefix no matter how the keys are crafted externally
+	// they won't interfiere with the database
+	//
+	// Having this commong prefix also also allows to store metadata that may be
+	// needed, without polluting the key spaces.
+	//
+	// Another side effect is that migrations are possible, by changing the
+	// prefix namespace, pulling the old prefix and migrating to the new one.
+	dataPrefix    = []byte("v1")
+	dataPrefixLen = len(dataPrefix)
+)
```

Comment:
Also, the decision on the order of key / inverse height is very interesting.

If key goes first, then reading the same key at different heights will be sequential reads.

If height goes first, then reading values modified at the same height will be more performant.

I expect our workload to primarily consist of a batch of read operations at a specific height. However, the read operations for a specific height may or may not tend towards reading values modified around the same height or be relatively random whether they were last modified at a much earlier height.

Reply:
> If height goes first, then reading values modified at the same height will be more performant.

This will only work if we know exactly at which height a change happened. Right now we can do a search such as a I know this `prefix` and I'd like to know the value at height 90, then an iterator is started requesting `prefix:90` _or the previous value_, which may be `prefix` at height 85 for example. If the key format shuffled, it would be impossible  to fulfil the most common usage case cheaply.